### ISSUE TITLE
while > 0 check on decrement doesn't register int as positive-int

### ISSUE THREAD

**Issue by bitwise-operators** [vimeo/psalm#6976]
When a `positive-int` var is modified using a decrement operator in a `while()` loop to check that it remains positive, Psalm still counts it as an `int`, no longer a `positive-int`. Which gives problems when using said variable in functions like `sleep()`, that require a `positive-int`.

https://psalm.dev/r/f8eed92abd

Also applies to do-while loops, or using the decrement operator inside the second argument of a for-loop.

**Comment by weirdan** [comment]
while the original bug remains (we infer `$a` to be `int<min, 2>` inside the loop) the snippet no longer emits any issues (I wonder why).

**Comment by bitwise-operators** [comment]
Interesting. Even adding a second parameter to the while-check doesn't matter, so it isn't that the code in the loop never gets touched.

I also checked, and `sleep()` still requires a `positive-int`. 

Did anything change in de code for that type?

**Comment by weirdan** [comment]
It's been 2 years, so there were quite a bit of changes. Looks like it affects internal functions only though: https://psalm.dev/r/e7de0789c5

**Comment by orklah** [comment]
Seems like sleep is the special case instead: https://psalm.dev/r/b0f7c4ff03

**Comment by weirdan** [comment]
No, it's not special. Its parameter type is just overridden in the stub: https://github.com/vimeo/psalm/blob/6b27a6d34bcb96296770b0b31508995de96b48a5/stubs/CoreGenericFunctions.phpstub#L1801-L1805

**Comment by bitwise-operators** [comment]
There's no docblock `@param` there. Does the stub itself override de type in the callmap?
Or does the absence of a `@param` with the stub create an automatic `@param int` from the function definition?

**Comment by weirdan** [comment]
As far as I recall, we ignore the function definition in the callmap entirely if there's a definition for the function in stubs.

**Comment by bitwise-operators** [comment]
Then this is probably a problem with more built-ins. 

One option would be to add all the callMap data to the stubs, but that'd be a maintenance nightmare. So changing that behaviour would probably be better, but I can't oversee how hard that would be to change.

What exactly is the purpose of the stubs?

**Comment by weirdan** [comment]
> What exactly is the purpose of the stubs?

To provide additional information the callmap format does not allow, e.g. taint flows, generics, etc.

An alternative to the automatic merging of data coming from stubs and the callmap could be a test ensuring the two provide the same signatures (#9843).
